homework6.js

```diff
@@ -0,0 +1,212 @@
+  this.name = name;
+};
+
+Penguin.prototype = new Bird();
+Penguin.prototype.move = function() { return 'swim' };
+Penguin.prototype.layEgg = function() { return new Egg(Penguin) };
+
+// Egg constructor
+var Egg = function(parentAnimal) {
+  this.parentAnimal = parentAnimal;
+};
+
+Egg.prototype.hatch = function(name) { return new this.parentAnimal(name) };
+
+// ---------
+// 3a)
+// ---------
+
+var Card = (function(){
+
+	// Constructor:
+	function Card(id) {
+		this.id = id;
+	}
+	var proto = Card.prototype; //create alias for convenience
+
+	// Instance methods (shared via prototype, inherited by subclass instances):
+	proto.isValid = function() {
+		return ((typeof this.id)==="number")//correct type
+			&& (this.id%1 === 0)  //integer
+			&& (this.id >=0)  // non-negative
+			&& (this.id < this.constructor.numCards()); //in range for card type
+	}
+
+	proto.rank = function() {
+		return Math.floor(this.id/4)+1;
+	}
+	proto.suit = function() {
+		return (this.id%4)+1;
+	}
+	proto.color = function() {
+		var suitVal=this.suit();
+        return suitVal && ((suitVal<3)? "red": "black");
+	}
+	proto.name = function() {
+        var rankVal = this.rank();
+        var suitVal = this.suit();
+        return rankVal && suitVal &&
+        	// Instead of referencing the private arrays specific to this class,
+        	// use the class method (of the instance's constructor)
+        	// so that a different array can be returned for each subclass
+            (this.constructor.rankNames()[rankVal]
+            +' of '
+            +this.constructor.suitNames()[suitVal]);
+	}
+
+	// Private data:
+	var suitNames = ['','Hearts','Diamonds','Spades','Clubs'];
+
+	var rankNames = ['','Ace','Two','Three','Four','Five','Six','Seven',
+					'Eight','Nine','Ten','Jack','Queen','King'];
+
+
+	// Class methods (not inherited by subclasses):
+	Card.isCard = function(thing) {
+		return (thing.constructor === this)
+			&& thing.isValid();
+	}
+
+	Card.rankNames = function() {
+		return rankNames.slice();
+	}
+
+	Card.suitNames = function() {
+		return suitNames.slice();
+	}
+
+	Card.numCards = function() {
+		return 52;
+	}
+
+	return Card;
+})();
+
+var TarotCard = (function() {
+  function TarotCard (id) {
+    this.id = id;
+  }
+
+  TarotCard.prototype = new Card();
+  TarotCard.prototype.constructor = TarotCard;
+
+
+	TarotCard.prototype.color = undefined;
+
+	// Private data:
+	var suitNames = ['','Cups','Pentacles','Swords','Wands'];
+
+	var rankNames = ['','Ace','Two','Three','Four','Five','Six','Seven',
+					'Eight','Nine','Ten','Page','Knight','Queen','King'];
+
+  TarotCard.isCard = function(thing) {
+		return (thing.constructor === this)
+			&& thing.isValid();
+	}
+
+	TarotCard.rankNames = function() {
+		return rankNames.slice();
+	}
+
+	TarotCard.suitNames = function() {
+		return suitNames.slice();
+	}
+
+  TarotCard.numCards = function() {
+		return 56;
+	}
+
+  return TarotCard;
+})();
+
+// ---------
+// 3b)
+// ---------
+
```

Comment:
3b isn't pulling card images for me, but this is largely because the images as still saved as titles, eg 10_of_clubs.svg, etc. I tried changing line 207 below to conform to pulling names but your ImageCard isn't able to use the function rankNames (which is an important part of the names method). Play around with your code a bit so that your ImageCard instance has access to this information and can then access its own name.